src/main/java/org/folio/service/s3storage/MinioStorageServiceImpl.java

```diff
@@ -91,160 +80,115 @@ public Future<FileUploadInfo> getFileUploadPartUrl(
         } catch (S3ClientException e) {
           blockingFuture.fail(e);
         }
-      },
-      (AsyncResult<String> asyncResult) -> {
-        if (asyncResult.failed()) {
-          promise.fail(asyncResult.cause());
-        } else {
-          String url = asyncResult.result();
-          FileUploadInfo fileUpload = new FileUploadInfo();
-          fileUpload.setUrl(url);
-          fileUpload.setKey(key);
-          fileUpload.setUploadId(uploadId);
-          promise.complete(fileUpload);
-        }
-      }
-    );
-
-    return promise.future();
+      })
+      .map(url ->
+        new FileUploadInfo().withUrl(url).withKey(key).withUploadId(uploadId)
+      );
   }
 
   @Override
   public Future<FileDownloadInfo> getFileDownloadUrl(String key) {
-    Promise<FileDownloadInfo> promise = Promise.promise();
     FolioS3Client client = folioS3ClientFactory.getFolioS3Client();
 
-    vertx.executeBlocking(
-      (Promise<String> blockingFuture) -> {
+    return vertx
+      .executeBlocking((Promise<String> blockingFuture) -> {
         try {
+          verifyKey(key);
+
           LOGGER.info("Getting presigned URL for key {}", key);
           blockingFuture.complete(client.getPresignedUrl(key));
         } catch (S3ClientException e) {
           blockingFuture.fail(e);
         }
-      },
-      (AsyncResult<String> asyncResult) -> {
-        if (asyncResult.failed()) {
-          promise.fail(asyncResult.cause());
-        } else {
-          promise.complete(
-            new FileDownloadInfo().withUrl(asyncResult.result())
-          );
-        }
-      }
-    );
-
-    return promise.future();
+      })
+      .map(url -> new FileDownloadInfo().withUrl(url));
   }
 
   @Override
   public Future<InputStream> readFile(String key) {
-    Promise<InputStream> inStreamPromise = Promise.promise();
     FolioS3Client client = folioS3ClientFactory.getFolioS3Client();
 
-    vertx.executeBlocking(
-      (Promise<InputStream> blockingFuture) -> {
-        try {
-          LOGGER.info(
-            "Created input stream to read remote file for key {}",
-            key
-          );
-          InputStream inStream = client.read(key);
-          blockingFuture.complete(inStream);
-        } catch (S3ClientException e) {
-          LOGGER.error("Could not read from S3:", e);
-          blockingFuture.fail(e);
-        }
-      },
-      (AsyncResult<InputStream> asyncResult) -> {
-        if (asyncResult.failed()) {
-          inStreamPromise.fail(asyncResult.cause());
-        } else {
-          inStreamPromise.complete(asyncResult.result());
-        }
+    return vertx.executeBlocking((Promise<InputStream> blockingFuture) -> {
+      try {
+        verifyKey(key);
+
+        LOGGER.info("Created input stream to read remote file for key {}", key);
+        InputStream inStream = client.read(key);
+        blockingFuture.complete(inStream);
+      } catch (S3ClientException e) {
+        LOGGER.error("Could not read from S3:", e);
```

Comment:
Is there an opportunity to log something else? I mean filename or some other identifier?

Reply:
It should contain detailed information in the exception; additionally, it logs right before `Created input stream to read remote file for key {}`, which would contain the identifier.